src/external/dftk_to_wannier90.jl

```diff
@@ -0,0 +1,605 @@
+            end
+        end
+        write(f,"end atoms_frac"*"\n"^2)
+
+        ## Projection bloc : message for the user.
+        write(f,"""
+        ! Add a projection bloc only with the option : guess = "win" in dftk2wan_wannierization_files.
+        ! Otherwise, DFTK uses by default gaussian guesses on specified centers 
+        ! or the SCDM method if guess = "SCDM" (not yet implemented).
+
+
+        """)
+
+
+        ## k points
+        write(f,"!"^20*" k_points"*"\n"^2)
+
+        # Bands_plot bloc
+        if bands_plot
+            trash,labels,path = high_symmetry_kpath(basis.model)
+            write(f,"begin kpoint_path"*"\n")
+            for i in 1:size(path[1],1)-1
+                ## write path section A -> B
+                A,B = path[1][i],path[1][i+1]
+                # kpoints are writen in the file with normalized coordinates
+                k_A = round.(DFTK.normalize_kpoint_coordinate.(get!(labels,A,1)), digits = 5)
+                k_B = round.(DFTK.normalize_kpoint_coordinate.(get!(labels,B,1)), digits = 5)
+                # Wannier90 convention for the Γ point is G
+                if A == "\\Gamma"
+                    A = "G"
+                end
+                if B == "\\Gamma"
+                    B = "G"
+                end
+                write(f,"$A $(k_A[1]) $(k_A[2]) $(k_A[3]) $B $(k_B[1]) $(k_B[2]) $(k_B[3])"*"\n") 
+            end
+            write(f,"end kpoint_path"*"\n")
+            write(f,"bands_plot = T"*"\n"^2)
+        end
+        
+        # Mp grid
+        if kgrid[1]*kgrid[2]*kgrid[3] !== size(scfres.ψ,1)
+            error("Given kgrid doesn't match the kgrid used for scf calculation")
+        end
+        write(f,"mp_grid : $(kgrid[1]) $(kgrid[2]) $(kgrid[3])"*"\n")
+        if kgrid == [1,1,1]
+            write(f,"gamma_only      = true"*"\n")
+        else
+            write(f,"\n")
+        end
+        
+        # kpoints bloc
+        write(f,"begin kpoints"*"\n")
+        for i in 1:size(scfres.ψ,1)
+            for x in basis.kpoints[i].coordinate
+                write(f,"$x ")
+            end
+            write(f,"\n")
+        end
+        write(f,"end kpoints"*"\n"^2)
+
+        @info "File generated.",f
+    end
+    
+    
+end
+
+
+"""
+Read the .nnkp file provided by the preprocessing routine of Wannier90 (i.e. "wannier90.x -pp prefix")
+Returns: 
+1) the array 'nn_kpts' of k points, they respective neirest neighbours and associated shifing vectors (non zero if the neighbour of is located in another cell).
+2) the number 'nn_num' of neighbours per k point. 
+2) the array 'projs' of projections. Each line contains the informations for one projection :  [center],[quantum numbers: l,mr,r],[ref. z_axis],[ref. x_axis],α = Z/a
+"""
+function read_nnkp_file(prefix::String)
+
+    ## Check the presence of the file
+    if !isfile("$prefix.nnkp")
+        error("file $prefix.nnkp not found.")   
+    end
+
+    ## Read the file
+    file = open("$prefix.nnkp")
```

Comment:
Don't do it like this, use the `do` ... `end` syntax, because this automatically closes the file once you are done. Also I would avoid reading the complete file into the `ln` array ... in case something goes wrong you might end up running out of memory or it just takes forever. I'd just try to loop over the file and extract the relevant blocks or process it on the fly.

Reply:
Not sure I agree here: yes in theory it's better to minimize passes over the data, but if it's easier to just read the whole thing to search within it, just do it like that. nnkp should be relatively small files (eg compared with mmn which are huge), so no issue here